src/tests/cwrap/common_mock_nss_dl_load.c

```diff
@@ -0,0 +1,115 @@
+/*
+    Authors:
+        Iker Pedrosa <ipedrosa@redhat.com>
+
+    Copyright (C) 2021 Red Hat
+
+    SSSD tests: Fake nss dl load
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <dlfcn.h>
+
+#include "common_mock_nss_dl_load.h"
+
+#define LIB_PATH    "libnss_wrapper.so"
+
+
+static errno_t load_function(const char *fname, void **fptr)
+{
+    void *dl_handle = NULL;
+    errno_t ret;
+
+    dl_handle = dlopen(LIB_PATH, RTLD_NOW);
+    if (dl_handle == NULL) {
+        DEBUG(SSSDBG_FATAL_FAILURE, "Unable to load %s module, "
+              "error: %s\n", LIB_PATH, dlerror());
+        ret = ELIBACC;
+        goto done;
+    }
+
+    *fptr = dlsym(dl_handle, fname);
+    if (*fptr == NULL) {
+        DEBUG(SSSDBG_FATAL_FAILURE, "Library '%s' did not provide "
+              "mandatory symbol '%s', error: %s.\n",
+              LIB_PATH, fname, dlerror());
+        ret = ELIBBAD;
+        goto done;
+    }
+
+    ret = EOK;
+
+done:
+    return ret;
+}
+
+static enum nss_status
+mock_getpwnam_r(const char *name, struct passwd *result,
+                char *buffer, size_t buflen, int *errnop)
+{
+    void *pwd_pointer = NULL;
+    int rc;
+    errno_t (*fptr)(const char *, struct passwd *, char *,
+                    size_t, struct passwd **) = NULL;
+
+    rc = load_function("getpwnam_r", (void *)&fptr);
```

Comment:
Do you actually need to load symbol manually?
`libnss_wrapper` is already preloaded.
Imo, you can just use `getpwnam()` here.

Reply:
That works but I'm lost as to why it does. I guess it's because `libnss_wrapper` is loaded from the testing environment and when the symbol `getpwnam_r()` is searched the one from `libnss_wrapper` is returned. Is my understanding correct?

Reply:
Yes: https://github.com/SSSD/sssd/blob/1cc7b8021aa39931c4294c476ed89fac5c1bb159/src/tests/cwrap/cwrap_test_setup.sh#L13

This is exactly why `usertools.c::sss_user_by_name_or_uid()` originally calling `getpwnam()` worked in the test.